arches_lingo/views/trees.py

```diff
@@ -0,0 +1,365 @@
+    FloatField,
+    F,
+    OuterRef,
+    Subquery,
+    Value,
+)
+from django.db.models.expressions import CombinedExpression, Func
+from django.utils.translation import gettext_lazy as _
+from django.utils.decorators import method_decorator
+from django.views.generic import View
+
+from arches.app.models.models import (
+    Language,
+    ResourceInstance,
+    TileModel,
+    Value as ConceptValue,
+)
+from arches.app.models.system_settings import settings
+from arches.app.utils.decorators import group_required
+from arches.app.utils.response import JSONResponse
+
+from arches_lingo.const import (
+    SCHEMES_GRAPH_ID,
+    TOP_CONCEPT_OF_NODE_AND_NODEGROUP,
+    CLASSIFICATION_STATUS_NODEGROUP,
+    CLASSIFICATION_STATUS_ASCRIBED_CLASSIFICATION_NODEID,
+    CONCEPT_NAME_NODEGROUP,
+    CONCEPT_NAME_CONTENT_NODE,
+    CONCEPT_NAME_LANGUAGE_NODE,
+    CONCEPT_NAME_TYPE_NODE,
+    HIDDEN_LABEL_VALUE_ID,
+    SCHEME_NAME_NODEGROUP,
+    SCHEME_NAME_CONTENT_NODE,
+    SCHEME_NAME_LANGUAGE_NODE,
+    SCHEME_NAME_TYPE_NODE,
+    PREF_LABEL_VALUE_ID,
+    ALT_LABEL_VALUE_ID,
+)
+from arches_lingo.models import VwLabelValue
+
+TOP_CONCEPT_OF_LOOKUP = f"data__{TOP_CONCEPT_OF_NODE_AND_NODEGROUP}"
+BROADER_LOOKUP = f"data__{CLASSIFICATION_STATUS_ASCRIBED_CLASSIFICATION_NODEID}"
+
+cache = caches["lingo"]
+
+
+class JsonbArrayElements(Func):
+    """https://forum.djangoproject.com/t/django-4-2-behavior-change-when-using-arrayagg-on-unnested-arrayfield-postgresql-specific/21547/5"""
+
+    arity = 1
+    contains_subquery = True
+    function = "JSONB_ARRAY_ELEMENTS"
+
+
+class LevenshteinLessEqual(Func):
+    arity = 3
+    function = "LEVENSHTEIN_LESS_EQUAL"
+
+
+@method_decorator(
+    group_required("RDM Administrator", raise_exception=True), name="dispatch"
+)
+class ConceptTreeView(View):
+    def __init__(self):
+        super().__init__()
+        self.schemes = ResourceInstance.objects.none()
+
+        # key=concept valueid (str) val=language code
+        self.language_concepts: dict[str:str] = {}
+        # key=scheme resourceid (str) val=set of concept resourceids (str)
+        self.top_concepts: dict[str : set[str]] = defaultdict(set)
+        # key=concept resourceid (str) val=set of concept resourceids (str)
+        self.narrower_concepts: dict[str : set[str]] = defaultdict(set)
+        # key=resourceid (str) val=list of label dicts
+        self.labels: dict[str : list[dict]] = defaultdict(set)
+
+        # Maps representing a reverse (leaf-first) tree
+        # key=resourceid (str) val=set of concept resourceids (str)
+        self.broader_concepts: dict[str : set[str]] = defaultdict(set)
+        # key=resourceid (str) val=set of scheme resourceids (str)
+        self.schemes_by_top_concept: dict[str : set[str]] = defaultdict(set)
+
+        self.read_from_cache()
+
+        # Not currently cached because written to during serialization.
+        self.polyhierarchical_concepts = set()
+
+    def read_from_cache(self):
```

Comment:
Caching isn't working on my local right now. I'll look into it a little bit, but we may also find we don't even need this. Would be nice to check this on realistic cloud infra with a realistic dataset.

Reply:
Culprit was a typo. Turns out you can never be in too much of a hurry to skip creating python constants 😅